### PR TITLE
Fix compilation warnings

### PR DESCRIPTION
**Pull request by kpadmasola**
+ steps followed:
+ uncomment -Wall in package.yaml
+ run hpack . to generate cabal file again
+ and then try cabal build
+ fix any warnings reported by GHC

### CODE REVIEW

File: lib/Eclair/AST/Lower.hs [hunk 1]
```diff
@@ -28,10 +28,10 @@ compileToRA ast = RA.Module $ concatMap processDecls sortedDecls where
 
   scc :: AST -> [[AST]]
   scc = \case
-    Module _ decls -> map G.flattenSCC sortedDecls
+    Module _ decls -> map G.flattenSCC sortedDecls'
       where
         relevantDecls = filter isRuleOrAtom decls
-        sortedDecls = G.stronglyConnComp $ zipWith (\i d -> (d, i, refersTo d)) [0..] relevantDecls
```
Comment (luc-tielen) [inline lib/Eclair/AST/Lower.hs:34]: Oh wow, this is a nasty case of shadowing. Good that this is fixed now.

(I checked this change myself manually just now, the typechecker would've caught it, but still..)

File: lib/Eclair/EIR/Codegen.hs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Eclair.EIR.Codegen
-  ( CodegenT(..)
+  ( CodegenT()
```
Comment (luc-tielen) [inline lib/Eclair/EIR/Codegen.hs:2]: Type synonyms usually are exported like below: 

```suggestion
  ( CodegenT
```

(Apply this suggestion to fix it.)

File: lib/Eclair/EIR/Lower.hs [hunk 3]
```diff
@@ -289,13 +285,6 @@ codegenSymbolTable exts = do
   where
     intoIO = pure . runIdentity
 
--- TODO: add hash based on filepath of the file we're compiling?
-mkType :: Name -> [Table] -> ModuleBuilderT IO LLVM.Type
```
Comment (luc-tielen) [inline lib/Eclair/EIR/Lower.hs:293]: This must've been a piece of code I forgot to delete during the addition of string support! 

Nice.

File: lib/Eclair/Error.hs [hunk 4]
```diff
@@ -251,11 +252,11 @@ style = reAnnotate style
...
       KindColor isError ->
-        bold <> style (ThisColor isError)
+        bold <> style' (ThisColor isError)
       NoLineColor ->
         bold <> colorDull Magenta
       MarkerStyle st ->
-        bold <> style st
+        bold <> style' st
       CodeStyle ->
         color White
```
Comment (luc-tielen) [inline lib/Eclair/Error.hs:262]: Many changes in this file.. but in the future we will probably swap this out for our own error rendering implementation anyway. :)

File: lib/Eclair/LLVM/BTree.hs [hunk 5]
```diff
@@ -170,18 +167,17 @@ generateTypes sizes = mdo
   pure $ Types
     { btreeTy = btreeTy
     , iteratorTy = btreeIteratorTy
-    , nodeSizeTy = nodeSizeTy
-    , nodeTypeTy = nodeTypeTy
+    , nodeSizeTy = nodeSizeTy'
+    , nodeTypeTy = nodeTypeTy'
     , nodeTy = nodeTy
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/BTree.hs:172]: Interesting that it doesn't warn on these recursively defined types. But it's fine, leave as-is.

File: lib/Eclair/LLVM/Codegen.hs [hunk 6]
```diff
@@ -19,7 +17,7 @@ import Eclair.LLVM.Config
 
 
 llvmSizeOf :: (MonadModuleBuilder m, MonadIO m)
-           => ForeignPtr LibLLVM.Context -> Ptr LibLLVM.TargetData -> Type -> m Word64
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/Codegen.hs:22]: This doesn't show up for me when I do it locally? Why did you add the fully qualified name?

Can we try removing it here (probably by changing the way we import modules in this file)?

File: lib/Eclair/LLVM/Codegen.hs [hunk 7]
```diff
@@ -32,23 +30,23 @@ withLLVMTypeInfo ctx m = do
...
+  _ <- liftIO $ M.traverseWithKey (serialize ctx) structTys
   -- Finally, we can call the function with all type info available in LLVM.
   m
   where
     forwardDeclareStruct ctx name structTy =
       (,structTy) <$> LibLLVM.mkOpaqueStructType ctx name
 
-    serialize :: ForeignPtr LibLLVM.Context -> Name -> (Ptr LibLLVM.Type, Type) -> IO ()
-    serialize ctx name (llvmTy, ty) = case ty of
+    serialize :: Foreign.ForeignPtr.ForeignPtr LibLLVM.Context -> Name -> (Ptr LibLLVM.Type, Type) -> IO ()
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/Codegen.hs:40]: Same comment here, I really try to avoid these fully qualified types.

File: lib/Eclair/LLVM/Codegen.hs [hunk 8]
```diff
@@ -32,23 +30,23 @@ withLLVMTypeInfo ctx m = do
...
+    serialize ctx _ (llvmTy, ty) = case ty of
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/Codegen.hs:41]: Remove the `_` entirely, and change `M.traverseWithKey` on line 33 to `traverse`

File: lib/Eclair/LLVM/Codegen.hs [hunk 9]
```diff
@@ -32,23 +30,23 @@ withLLVMTypeInfo ctx m = do
...
       StructureType packed tys -> do
         tys' <- traverse (encodeType ctx) tys
         LibLLVM.setNamedStructBody llvmTy tys' packed
-      ty ->
+      _ ->
         panic $ "Unexpected typedef: only structs are allowed, but got: " <> show ty
 
 -- NOTE: this only works if all the named structs are known beforehand (a.k.a. forward declared)!
-encodeType :: ForeignPtr LibLLVM.Context -> Type -> IO (Ptr LibLLVM.Type)
+encodeType :: Foreign.ForeignPtr.ForeignPtr LibLLVM.Context -> Type -> IO (Ptr LibLLVM.Type)
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/Codegen.hs:49]: Same comment about fully qualified type

File: lib/Eclair/LLVM/HashMap.hs [hunk 10]
```diff
@@ -35,7 +35,7 @@ data HashMap
 
 data CGState
   = CGState
-  { externals :: Externals
+  { _externals :: Externals
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/HashMap.hs:38]: I checked in the code, this field can be removed entirely. It only needs to passed into the `Vector.codegen` function.

File: lib/Eclair/LLVM/SymbolTable.hs [hunk 11]
```diff
@@ -23,23 +23,22 @@ data SymbolTable
 data CGState
   = CGState
   { symbolTableTy :: Type
-  , symbolTy :: Type
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/SymbolTable.hs:26]: I don't like that we need to rename our field names just to silence a warning. Can we turn off unused warning for this file? (With a `{-# OPTIONS_GHC ... #-}`

File: lib/Eclair/LLVM/Vector.hs [hunk 12]
```diff
@@ -34,10 +34,10 @@ type Destructor = Operand -> IRCodegen ()
 
 data CGState
   = CGState
-  { externals :: Externals
```
Comment (luc-tielen) [inline lib/Eclair/LLVM/Vector.hs:37]: Same comment here as in symbol table, revert back to the old names and turn off the specific warning for this module.

File: lib/Eclair/Literal.hs [hunk 13]
```diff
@@ -1,9 +1,16 @@
...
   = LNumber Word32
   | LString Text
   deriving (Eq, Show)
 
+instance Pretty Literal where
+  pretty = \case
+    LNumber x -> pretty x
+    LString x -> pretty x
+
+
```
Comment (luc-tielen) [inline lib/Eclair/Literal.hs:16]:
```suggestion
```

(Try to never include trailing whitespace in a module.)

File: lib/Eclair/RA/Codegen.hs [hunk 14]
```diff
@@ -288,15 +286,15 @@ withUpdatedAlias :: Alias -> EIR -> CodegenM a -> CodegenM a
...
   local (const state') m
   where
-    updateAlias ls curr =
-      ls { aliasMap = M.insert a curr (aliasMap ls) }
+    updateAlias ls curr' =
+      ls { aliasMap = M.insert a curr' (aliasMap ls) }
 
 withEndLabel :: EIR.LabelId -> CodegenM a -> CodegenM a
 withEndLabel end m = do
-  state <- ask
```
Comment (luc-tielen) [inline lib/Eclair/RA/Codegen.hs:299]: Remove this line entirely, this is dead code.

File: lib/Eclair/TypeSystem.hs [hunk 15]
```diff
@@ -153,8 +150,7 @@ checkDecls :: TypedefMap -> AST -> [TypeError]
 checkDecls typedefMap = \case
   Module _ decls ->
     -- NOTE: By doing runM once per decl, each decl is checked with a clean state
-    let state = CheckState typedefMap mempty mempty
```
Comment (luc-tielen) [inline lib/Eclair/TypeSystem.hs:156]: Dang, can't believe I left this one around! :sweat_smile:

File: package.yaml [hunk 16]
```diff
@@ -72,7 +72,7 @@ default-extensions:
   - PatternSynonyms
 
 ghc-options:
-  #- -Wall TODO turn on again
+  - -Wall # TODO turn on again
```
Comment (luc-tielen) [inline package.yaml:75]: Remove comment
```suggestion
  - -Wall
```

**Review by luc-tielen — CHANGES_REQUESTED** [review]
Wow, so many improvements; this must've been a large effort!
Thanks for doing this.

I left a few comments, but in general it's looking great! If you fix the last few remarks I will merge it in. :smile:

File: lib/Eclair/Error.hs (hunk 4 above)
Comment (kpadmasola) [reply lib/Eclair/Error.hs:262]: one of the key changes is the line
```
{-# OPTIONS_GHC -fno-warn-orphans #-}
```
at the top. I am not sure if that is the appropriate fix or if there is another more appropriate alternative.

File: lib/Eclair/LLVM/Codegen.hs (hunk 6 above)
Comment (kpadmasola) [reply lib/Eclair/LLVM/Codegen.hs:22]: I think I may have selected the less appropriate option in the IDE when applying the fix. Will remove.

**Comment by kpadmasola** [comment]
@luc-tielen thanks for the review. I applied the suggested changes and pushed again.

File: lib/Eclair/Error.hs (hunk 4 above)
Comment (luc-tielen) [reply lib/Eclair/Error.hs:262]: I saw that, but it's fine. Soon we will replace most of this module anyway.

File: .devcontainer/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,9 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.245.2/containers/alpine/.devcontainer/base.Dockerfile
```
Comment (luc-tielen) [inline .devcontainer/Dockerfile:1]: What is this file? If something for local development, remove and add to gitignore

File: .devcontainer/devcontainer.json [hunk 18]
```diff
@@ -0,0 +1,22 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
```
Comment (luc-tielen) [inline .devcontainer/devcontainer.json:1]: Remove

File: out [hunk 19]
```diff
@@ -0,0 +1,72 @@
+Cloning into '/Users/padmasol/fun/github/placidex/eclair-lang/dist-newstyle/src/diagnose-2c684c8ab7fd229a'...
```
Comment (luc-tielen) [inline out:1]: Remove

**Review by luc-tielen — CHANGES_REQUESTED** [review]
Remove dev container files and then we can merge!

File: .devcontainer/Dockerfile (hunk 17 above)
Comment (kpadmasola) [reply .devcontainer/Dockerfile:1]: sorry about that, will remove.

**Comment by kpadmasola** [comment]
@luc-tielen hopefully I have fixed all the issues from the code review now 😅

File: out (hunk 19 above)
Comment (kpadmasola) [reply out]: done

File: .devcontainer/devcontainer.json (hunk 18 above)
Comment (kpadmasola) [reply .devcontainer/devcontainer.json]: done

File: .devcontainer/Dockerfile (hunk 17 above)
Comment (kpadmasola) [reply .devcontainer/Dockerfile]: done

**Comment by kpadmasola** [comment]
> Remove dev container files and then we can merge!

done!

**Review by luc-tielen — APPROVED** [review]

**Comment by luc-tielen** [comment]
Thanks!
